### PR TITLE
Crs 1751 endpoints for discrepancy form

### PR DESCRIPTION
**Pull request by justin-gardiner**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt [hunk 1]
```diff
@@ -101,16 +111,74 @@ class ComparisonService(
...
   }
 
+  fun createDiscrepancy(
+    comparisonReference: String,
+    comparisonPersonReference: String,
+    discrepancyRequest: CreateComparisonDiscrepancyRequest,
+  ): ComparisonDiscrepancySummary {
+    log.info("creating discrepancy $discrepancyRequest for comparison $comparisonReference and comparison person $comparisonPersonReference")
+    val comparison = comparisonRepository.findByComparisonShortReference(comparisonReference)
+      ?: throw EntityNotFoundException("Could not find comparison with reference: $comparisonReference ")
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt:132]: Shouldn't we be preventing users who don't have access to the prison for which the mismatch exists from creating a discrepancy?

I get that they'd not be able to see it in the first instance, it's just protecting ourselves 🤷‍♂️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt [hunk 2]
```diff
@@ -101,16 +111,74 @@ class ComparisonService(
...
       val overrideDates = objectMapper.convertValue(comparisonPerson.overrideDates, object : TypeReference<Map<ReleaseDateType, LocalDate?>>() {})
       val breakdownByReleaseDateType = objectMapper.convertValue(comparisonPerson.breakdownByReleaseDateType, object : TypeReference<Map<ReleaseDateType, ReleaseDateCalculationBreakdown>>() {})
       val sdsPlusSentences = if (comparisonPerson.sdsPlusSentencesIdentified.isEmpty) emptyList<SentenceAndOffences>() else objectMapper.convertValue(comparisonPerson.sdsPlusSentencesIdentified, object : TypeReference<List<SentenceAndOffences>>() {})
-      return transform(comparisonPerson, nomisDates, calculatedReleaseDates, overrideDates, breakdownByReleaseDateType, sdsPlusSentences)
+      return transform(comparisonPerson, nomisDates, calculatedReleaseDates, overrideDates, breakdownByReleaseDateType, sdsPlusSentences, hasDiscrepancyRecorded)
     }
     throw CrdWebException("Forbidden", HttpStatus.FORBIDDEN, 403.toString())
   }
 
+  fun createDiscrepancy(
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt:125]: Q: Should this be transactional?
As in if we can't record the causes don't persist it?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt [hunk 3]
```diff
@@ -147,6 +150,58 @@ class ComparisonController(
     return comparisonService.getComparisonPersonByShortReference(comparisonReference, mismatchReference)
   }
 
+  @PostMapping(value = ["{comparisonReference}/mismatch/{mismatchReference}/discrepancy"])
+  @PreAuthorize("hasAnyRole('ROLE_RELEASE_DATE_COMPARER')")
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt:154]: Should this include `ROLE_RELEASE_DATE_MANUAL_COMPARER`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt [hunk 4]
```diff
@@ -147,6 +150,58 @@ class ComparisonController(
...
+    @Valid
+    @RequestBody
+    discrepancyInput: CreateComparisonDiscrepancyRequest,
+  ): ComparisonDiscrepancySummary {
+    log.info("Request received to create a new discrepancy -- $discrepancyInput")
+    return comparisonService.createDiscrepancy(comparisonReference, mismatchReference, discrepancyInput)
+  }
+
+  @GetMapping(value = ["{comparisonReference}/mismatch/{mismatchReference}/discrepancy"])
+  @PreAuthorize("hasAnyRole('ROLE_RELEASE_DATE_COMPARER')")
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt:182]: Should this include ROLE_RELEASE_DATE_MANUAL_COMPARER?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt (hunk 2 above)
Comment (justin-gardiner) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt:125]: yes good spot will fix

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt (hunk 3 above)
Comment (justin-gardiner) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt:154]: Yes I've added the manual comparison version of these endpoints and added the role there.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt (hunk 4 above)
Comment (justin-gardiner) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/resource/ComparisonController.kt:182]: as above, I've added to ManualComparisonController.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt (hunk 1 above)
Comment (justin-gardiner) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt:125]: added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt (hunk 2 above)
Comment (justin-gardiner) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ComparisonService.kt:118]: added to BulkComparisonService.createDiscrepancy which is called by ComparisonService and ManualComparisonService

**Review by srowlands-moj — DISMISSED** [review]
LGTM - if the cause categories become something that is frequently changed might want to consider making these reference data in a new DB table.

**Review by srowlands-moj — APPROVED** [review]

**Pull request merged**